### PR TITLE
Adding `custom` colors to the design system

### PR DESCRIPTION
**Pull request by ckark**
# What does this PR do?

Part of cleaning up dark mode requires adding a handful of color styles to the palette. In order for this not to be disruptive, and given MUI's many limitations, I've created a `custom` palette that standardizes background, text, and border styles.

An advantage of this approach is that the new colors won't appear unless explicitly called. It will drastically cut down on the use of `isDark` in ternary operations used through the `grid` monorepo.

What follows is a visual overview of the changes.

## Background
image
## Text
image
## Borders
image

Additional screens that depict the difference appear in [Figma](https://www.figma.com/file/GZwU6yZbsPWXVhyl9TmaPN/Dark-Mode?type=design&node-id=0%3A1&mode=design&t=jo7QeNO7nLu5aZrh-1).

## Limitations

Declaring custom styles in MUI is cumbersome and requires extensive reworking of the type declarations in the design system—an undertaking worth considering, but well beyond the scope of this project.

## Test Plan

This PR is a cleaned up version of a few other draft PRs whose hash we copied into the `package.json` in the `grid` repo to test.

### CODE REVIEW

File: src/design-system/theme/palette.ts [hunk 1]
```diff
@@ -110,7 +110,27 @@ const error: PaletteColorOptions & Record<string, string> = {
   "contrastText": "#821D1E",
 };
 
+const customColors: PaletteColorOptions & Record<string, string> = {
```
Comment (yurijmikhalevich) [inline src/design-system/theme/palette.ts:113]: I like that we are adding these colors without changing everything right away, but I feel that if we don't finish the migration, we will end up with even more colors to manage.

Comment (ckark) [reply src/design-system/theme/palette.ts:113]: For sure. Can you elaborate what you mean by finishing the migration?

File: src/design-system/theme/paletteDark.ts [hunk 2]
```diff
@@ -109,11 +109,32 @@ const error: PaletteColorOptions & Record<string, string> = {
...
+  linkText: "#D0B6F6",
+  contrastText: "#1C1C1C",
+  primaryBorder: "#292929",
+  secondaryBorder: "#454545",
+  tertiaryBorder: "#4F4F4F",
+  quaternaryBorder: "#5C5C5C",
+};
+
 const grey: ColorPartial & Record<string, string> = {
+  ...customColors,
```
Comment (yurijmikhalevich) [inline src/design-system/theme/paletteDark.ts:133]: Why are these added to `grey` and to the `palette` object? In way of `palette.customColors`? Having `palette.grey.primaryText` not being grey is super confusing.

File: src/design-system/theme/paletteDark.ts [hunk 3]
```diff
@@ -109,11 +109,32 @@ const error: PaletteColorOptions & Record<string, string> = {
...
   "main": "#DEDFE3",
   "5": "#363636",
   "10": "#030303",
-  "20": "#363636",
+  "20": "#1C1C1C",
```
Comment (yurijmikhalevich) [inline src/design-system/theme/paletteDark.ts:137]: Can you add screenshots showing the impact of this change?

File: src/design-system/theme/palette.ts [hunk 4]
```diff
@@ -140,7 +160,7 @@ const text: Partial<TypeText> = {
   disabled: "#828282",
 };
 
-const divider = grey["20"];
+const divider = "#E0E0E0";
```
Comment (yurijmikhalevich) [inline src/design-system/theme/palette.ts:163]: Ditto.

File: src/design-system/theme/paletteDark.ts [hunk 5]
```diff
@@ -138,6 +143,36 @@ const text: Partial<TypeText> = {
...
 
 const divider = grey["20"];
+
+type ExtendedPaletteColorOptions = PaletteColorOptions & {
+  [key: string]: string | any;
+};
+
+declare module "@mui/material/styles/createPalette" {
+  export interface PaletteOptions {
+    custom?: ExtendedPaletteColorOptions;
```
Comment (yurijmikhalevich) [inline src/design-system/theme/paletteDark.ts:153]: This way it should be able to autocomplete and check for typos in `custom` palette name fields.

```suggestion
    custom?: typeof custom;
```

File: src/design-system/theme/palette.ts [hunk 6]
```diff
@@ -141,6 +145,36 @@ const text: Partial<TypeText> = {
...
+  textTertiary: "#696969",
+  textMuted: "#9C9C9C",
+  textDisabled: "#B5B5B5",
+  textLink: "#4F00BA",
+  textContrast: "#FFFFFF",
+  borderPrimary: "#E0E0E0",
+  borderSecondary: "#DEDEDE",
+  borderTertiary: "#D6D6D6",
+  borderQuaternary: "#CFCFCF",
+};
```
Comment (yurijmikhalevich) [inline src/design-system/theme/palette.ts:176]: Does this work?

```suggestion
const custom = {
  backgroundPrimary: "#FFFFFF",
  backgroundSecondary: "#FCFCFD",
  backgroundTertiary: "#F9FAFB",
  backgroundQuaternary: "#F2F4F7",
  backgroundContrast: "#292929",
  textPrimary: "#1C1C1C",
  textSecondary: "#4F4F4F",
  textTertiary: "#696969",
  textMuted: "#9C9C9C",
  textDisabled: "#B5B5B5",
  textLink: "#4F00BA",
  textContrast: "#FFFFFF",
  borderPrimary: "#E0E0E0",
  borderSecondary: "#DEDEDE",
  borderTertiary: "#D6D6D6",
  borderQuaternary: "#CFCFCF",
};

declare module "@mui/material/styles/createPalette" {
  export interface PaletteOptions {
    custom?: typeof custom;
  }
}
```

You may also need to say (will need to import `MuiPaletteOptions`:
```js
export interface PaletteOptions implements MuiPaletteOptions {
```
